### ISSUE TITLE
[BUG] Deepspeed ZeRO Optimizer Stage 2 Memory Leak? How to cleanup?

### ISSUE THREAD

**Issue by chiragjn** [microsoft/DeepSpeed#4856]
**Describe the bug**
I am using Huggingface transfomers Trainer with accelerate and have a general-purpose LM training script that looks like

```python
def train(training_arguments):
    tokenizer = ...
    model = AutoModelForCausalLM.from_pretrained(...)
    model = get_peft_model(model, lora_config)
    trainer = HFTrainer(
        model=model,
        tokenizer=tokenizer,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        args=training_arguments,
        data_collator=...,
        callbacks=...,
    )
    trainer.train(...)
    trainer.save_model(...)

def _cleanup_gpus():
    import gc, time, torch
    logger.info("Clearing gpus before moving ahead ...")
    for _ in range(6):
        gc.collect()
        time.sleep(10)
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
            torch.cuda.synchronize()
        logger.info(get_gpu_metrics())

def main():
    training_arguments = parser.parse_args_into_dataclasses()
    train(training_arguments)
    _cleanup_gpus()
   training_arguments.distributed_state.wait_for_everyone()
   merge_lora_adapters(...)  # This loads the model again in bf16 and merges adapters. I need GPU memory for this
```

Now when I run this script with DDP (`accelerate launch train.py ...`), `_cleanup_gpus` works as expected but when I use Deepspeed Zero 2 (`accelerate launch --use_deepspeed train.py --deepspeed config.json ...`) the GPU memory does not clear.

```json
{
  "bf16": {
    "enabled": true
  },
  "optimizer": {
    "type": "AdamW",
    "params": {
      "lr": "auto",
      "betas": "auto",
      "eps": "auto",
      "weight_decay": "auto"
    }
  },
  "zero_optimization": {
    "stage": 2,
    "overlap_comm": true,
    "contiguous_gradients": true,
    "sub_group_size": 1e9,
    "reduce_bucket_size": "auto",
    "allgather_partitions": true,
    "allgather_bucket_size": 5e8,
    "reduce_scatter": true,
    "round_robin_gradients": true
  },
  "gradient_accumulation_steps": "auto",
  "gradient_clipping": "auto",
  "steps_per_print": 2000,
  "train_batch_size": "auto",
  "train_micro_batch_size_per_gpu": "auto",
  "wall_clock_breakdown": false
}
```

Naturally, this means there are still references to the model parameters/optimizer somewhere, so I created weak references to all parameters in my `train()` function

```python
_param_weak_refs = []
def _setup_weak_refs(model):
    global _param_weak_refs
    for name, param in model.named_parameters():
        _param_weak_refs.append((name, weakref.ref(param)))
```

And after attaching a debugger just after `_cleanup_gpus()`  I can see `deepspeed.utils.debug.param_names` holds the parameter in a global instance
I cleared them and tried again

```python
def main():
    training_arguments = parser.parse_args_into_dataclasses()
    train(training_arguments)
    if training_arguments.deepspeed:
        deepspeed.utils.debug.module_names = {}
        deepspeed.utils.debug.param_names = {}
    _cleanup_gpus()
   training_arguments.distributed_state.wait_for_everyone()
   merge_lora_adapters(...)  # This loads the model again in bf16 and merges adapters
```

While the reference count went down it did not help, investigating again with a debugger I found DeepSpeedZeroOptimizer is still hanging around in memory

![sample-backref-graph](https://github.com/microsoft/DeepSpeed/assets/10295418/8a0aa6a6-c056-43cb-b3f8-5b01092b40f0)
![sample-backref-graph](https://github.com/microsoft/DeepSpeed/assets/10295418/ef25aa20-b9cb-436b-aeb4-7f908438fdc3)
The above is from a parameter that has requires_grad False
For requires_grad True the reference graph looks like this:
![sample-backref-graph](https://github.com/microsoft/DeepSpeed/assets/10295418/526d73ca-2dcb-4db3-b07d-16b4aea5779f)

I don't have much idea at the moment, I am trying to dig deeper, it seems to point to functions in `create_reduce_and_remove_grad_hooks`

**To Reproduce**
Train a model with Deepspeed Zero 2 config and try to clear gpu memory after finishing training. 
Please give me some time to make a cleaner standalone reproducible example

**Expected behavior**
Some way to completely destroy the deepspeed engine and clear gpu memory

**ds_report output**
```
[2023-12-21 05:15:48,112] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.1
 [WARNING]  using untested triton version (2.1.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/data/v/ft/lib/python3.10/site-packages/torch']
torch version .................... 2.1.1+cu118
deepspeed install path ........... ['/data/v/ft/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.12.5, unknown, unknown
torch cuda version ............... 11.8
torch hip version ................ None
nvcc version ..................... 11.8
deepspeed wheel compiled w. ...... torch 2.1, cuda 11.8
shared memory (/dev/shm) size .... 216.48 GB
```

**Screenshots**
NA

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 18.04]: Ubuntu 22.04
 - GPU count and types [e.g. two machines with x8 A100s each]: 1 machine with 4 x A10 GPUs (each card is 24 GB)
 - Interconnects (if applicable) [e.g., two machines connected with 100 Gbps IB]
 - Python version: 3.10
 - Any other relevant info about your setup

**Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?
I am using `accelerate launch`

**Docker context**
None

**Additional context**
NA

**Comment by chiragjn** [comment]
Also apologies, I filed the issue under `compression` by mistake, I intended it under `training`
I am not able to edit labels

**Comment by paraGONG** [comment]
Hello, I met the same problem. I am trying to call the train() function in a loop. I find each time it is called, the gpu memory will double. It seems that the previous memory is not released properly. Here is the detailed description: https://stackoverflow.com/questions/76186529/how-to-delete-memory-taken-up-by-deepspeed-engine
Have you solved the problem now?

**Comment by chiragjn** [comment]
@paraGONG I have raised a PR that works: https://github.com/microsoft/DeepSpeed/pull/4858
Waiting on authors for review

**Comment by paraGONG** [comment]
@chiragjn Thanks for your reply! I tried to call .destroy() on my deepspeed engine, but nothing happend. 
I read the source code of destroy(): https://github.com/microsoft/DeepSpeed/blob/827e1ca8aa76926d0765743efc59d8c2d1cb9db5/deepspeed/runtime/engine.py#L362C14-L362C14
`    def destroy(self):
        if self.optimizer is not None and hasattr(self.optimizer, 'destroy'):
            self.optimizer.destroy()`
then I found in my code hasattr(model.optimizer, 'destroy')=False
I use **FusedAdam** as my optimizer, however it does not have 'destroy' attribute.

**Comment by vergilus** [comment]
> [@paraGONG](https://github.com/paraGONG) I have raised a PR that works: [#4858](https://github.com/microsoft/DeepSpeed/pull/4858) With that PR you can call `.destroy()` on the deepspeed engine instance and after that `gc.collect()` and `torch.cuda.empty_cache()` should get the memory back Waiting on authors for review

I also ran into this issue lately. It went on fine with lora+deepspeed zero3, however, zero2 can not release the GPU memory. 
 There must be something wrong with the model parameter allocation.

**Comment by vergilus** [comment]
My current solution: update to 0.15.2, actively launch the `trainer.accelerator.free_memory()`, then `del trainer`.
